main.go

```diff
@@ -139,6 +143,52 @@ func HandleCard(uid []byte) (Result, error) {
 	return PaymentMade, nil
 }
 
+// RegisterUser creates a new row in the user table, with the given username
+// and password. It returns a populated User and no error on success. If the
+// username is already taken, it returns ErrUserExists.
+func RegisterUser(name string, password []byte) (*User, error) {
+	log.Printf("Registering user %s", name)
+
+	tx, err := db.Beginx()
+	if err != nil {
+		return nil, err
+	}
+	defer tx.Rollback()
+
+	pwhash, err := bcrypt.GenerateFromPassword(password, bcrypt.DefaultCost)
+	if err != nil {
+		return nil, err
+	}
+
+	// We need to check first if the username is already taken, because the
+	// error from an insert can't be checked programmatically.
+	var user User
+	if err := tx.Get(&user, `SELECT user_id, name, password FROM users WHERE name = $1`, name); err == nil {
+		return nil, ErrUserExists
+	}
+
+	result, err := tx.Exec(`INSERT INTO users (name, password) VALUES ($1, $2)`, name, pwhash)
+	if err != nil {
+		return nil, err
+	}
+
+	id, err := result.LastInsertId()
+	if err != nil {
+		if err := tx.Get(&id, `SELECT user_id FROM users WHERE name = $1`, name); err != nil {
+			return nil, err
+		}
+	}
```

Comment:
Is this something that can occur? LastInsertId() returning an error, but the id being correctly set nevertheless?